haskell: tune generic-builder for more parallel build performance

peti opened the pull request:
Raise the maximum number of cores used for parallel building from 4 to 16. Increase the size of the allocation area for GHC's garbage collector from 1 MiB to 64 MiB.

See https://www.twitch.tv/videos/611899011 for the motivation for this change.

nh2 reviewed (commented):
This looks sensible to me, but one question:

If it is done at the `ghcWithPackages` level, do/can/will we also have the ability to override it per Haskell package? I am suspecting that there might be certain large packages for which we may need to tune down `-j` or `-A` to not consume too much memory on builders that have less than the casual 3 TB RAM.

cdepillabout commented on pkgs/development/haskell-modules/with-packages-wrapper.nix:
```diff
@@ -2,6 +2,7 @@
 , withLLVM ? false
 , postBuild ? ""
 , ghcLibdir ? null # only used by ghcjs, when resolving plugins
+, ghcOptions ? []
```
Could you add a comment for this line, ideally with an example of what can be passed here?  I think an example like `["-j16" "+RTS" "-A64M" "-RTS"]` makes it immediately obvious to Haskellers what these options are.

peti commented:
It just occurred to me that the change to the ghcWithPackages wrapper is unnecessary, because you can do the same thing in `~/.cabal/config`. Adding 

~~~
program-default-options
  ghc-options: -j16 +RTS -A64M -RTS
~~~~

there has the same effect -- at least if you're building with cabal-install. I'm pretty sure an identical mechanism exists for `stack`, too. So, on second thought I'd rather not introduce yet another location to configure default flags that's specific to Nixpkgs.

peti committed:
haskell: tune generic-builder for more parallel build performance

Raise the maximum number of cores used for parallel building from 4 to 16.
Increase the size of the allocation area for GHC's garbage collector from 1 MiB
to 64 MiB.

See https://www.twitch.tv/videos/611899011 for the motivation for this change.

peti commented:
I'll merge this change to `haskell-updates` so that we can get some test results from Hydra before the live stream tomorrow.

expipiplus1 commented:
This change causes an uncomfortably high load on my poor machine.

![image](https://user-images.githubusercontent.com/857308/84470672-6d89e380-acb6-11ea-8289-c075ef0ae092.png)

I've also noticed loads of 120+ on my 8(16) core desktop, surely this can't be faster.

(sorry if I posted this before somewhere, I thought I did for another machine but couldn't find anything)
